pulsar/crypto/message_metadata.go

```diff
@@ -0,0 +1,122 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package crypto
+
+import (
+	pb "github.com/apache/pulsar-client-go/pulsar/internal/pulsar_proto"
+)
+
+// MessageMetadataSupplier wrapper implementation around message metadata
+type MessageMetadataSupplier interface {
+	// GetEncryptionKeys read all the encryption keys from the MessageMetadata
+	GetEncryptionKeys() []EncryptionKeyInfo
+
+	// UpsertEncryptionkey add new or update existing EncryptionKeys in to the MessageMetadata
+	UpsertEncryptionkey(EncryptionKeyInfo)
+
+	// GetEncryptionParam read the ecryption parameter from the MessageMetadata
+	GetEncryptionParam() []byte
+
+	// SetEncryptionParam set encryption parameter in to the MessageMetadata
+	SetEncryptionParam([]byte)
```

Comment:
Are setters need in this interface?

Reply:
Yes they are needed.
1. Producer will update `MessageMetadata` with encryption parameter and encrypted data key.
2. Consumer will try to read the same info from the `MessageMetadata` to decrypt the payload.

There is need to have this setters here since we are not directly exposing `proto package`.

 If client(user) wants to have his own implementation of `MessageCrypto` interface, in such case he won't be able to access `MessageMetadata` directly because it is an `internal` package. 

So adding this interface and exposing necessary methods to read `encrypted data key and encryption parameter` from the `MessageMetadata` that are required for decryption of payload.

Reply:
Ok makes sense. For my understanding what's the difference between `UpsertEncryptionkey` and `SetEncryptionParam` and when would one or the other be used?